tools/cache_testing.py

```diff
@@ -0,0 +1,116 @@
+        # Generate an array of uniform 0-1 pseudo-random values:
+        u = np.random.random(count)
+
+        # bisect them with distMap
+        v = np.searchsorted(distMap, u)
+
+        samples = [t-1 for t in v]
+
+        for sample in samples:
+            yield sample
+    else:
+        current_count = 0
+        while current_count < count:
+            # Generate an array of uniform 0-1 pseudo-random values, of the pipeline length:
+            u = np.random.random(pipeline)
+
+            # bisect them with distMap
+            v = np.searchsorted(distMap, u)
+
+            samples = [t-1 for t in v]
+            yield samples
+
+            current_count += len(samples)
+
+def update_stats(r, hits, misses, value_index, total_count):
+    """
+    A void function that uses terminal control sequences
+    to update hit/miss ratio stats for the user
+    while the testing tool runs.
+    """
+    percent_complete = (value_index + 1) / total_count
+
+    # Use the terminal control sequence to move the cursor to the beginning of the line
+    print("\r", end="")
+
+    # Print the loading bar and current hit rate
+    print("[{}{}] {:.0f}%, current hit rate: {:.6f}%".format("#" * int(percent_complete * 20), " " * int(20 - percent_complete * 20), percent_complete * 100, (hits / (hits + misses)) * 100), end="")
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Cache Benchmark', formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument('-c', '--count', type=int, default=100000, help='total number of incrby operations')
+    parser.add_argument('-u', '--uri', type=str, default='redis://localhost:6379', help='Redis server URI')
+    parser.add_argument('-a', '--alpha', type=int, default=1.0, help='alpha value being used for the Zipf distribution')
+    parser.add_argument('-n', '--number', type=int, default=30, help='the number of values to be used in the distribution')
+    parser.add_argument('-d', '--length', type=int, default=10, help='the length of the values to be used in the distribution')
+    parser.add_argument('-p', '--pipeline', type=int, default=0, help='pipeline size')
+
+    args = parser.parse_args()
+    uri = urlparse(args.uri)
+
+    r = redis.StrictRedis(host=uri.hostname, port=uri.port)
+
+    misses = 0
+    hits = 0
+
+    distribution_values_generator = rand_zipf_generator(args.number, args.alpha, args.count, args.pipeline)
+
+    if args.pipeline == 0:
+        for idx, val in enumerate(distribution_values_generator):
+            result = r.set(str(val), 'x' * args.length, nx=True)
+            if result:
+                misses += 1
+            else:
+                hits += 1
+            if idx % 50 == 0:
+                update_stats(r, hits, misses, idx, args.count)
+    else:
+        total_count = 0
+        for idx, values in enumerate(distribution_values_generator):
+            total_count += len(values)
+            p = r.pipeline()
```

Comment:
add `transaction=False`

Reply:
Oh wow, that sped things up significantly. Thanks!